File: packages/case-boundary/src/boundary/StateHandler.ts
import { CaseConfigurationError } from '@contract-case/case-core';
import { BoundaryResult } from './Result';

/**
 * An interface for a state handler that only has setup
 *
 * @public
 */
export abstract class BoundaryStateHandler {
  /**
   * Call the user's state setup function.
   *
   * If the user provided no setup function, this should be a function that does nothing and returns a `BoundarySuccess`
   *
   * @returns Either a `BoundaryFailure` with
   * `kind=BoundaryFailureKindConstants.CASE_CONFIGURATION_ERROR` or a
   * `BoundarySuccess` when there's no variables returned, or a
   * `BoundarySuccessWithMap` when there are variables returned
   */
  setup(): Promise<BoundaryResult> {
    throw new CaseConfigurationError(
      `${this}: State handler setup function not overridden`,
    );
  }

  /**
   * Call the user's state teardown function
   *
   * If the user provided no teardown function, this should be a function that does nothing and returns a `BoundarySuccess`
   *
   * @returns Either a `BoundaryFailure` with `kind=BoundaryFailureKindConstants.CASE_CONFIGURATION_ERROR` or a `BoundarySuccess`
   */
  teardown(): Promise<BoundaryResult> {
    throw new CaseConfigurationError(
      `${this}: State handler teardown function not overridden`,
    );
  }
}

File: packages/dsl-java/src/main/java/io/contract_testing/contractcase/configuration/StateHandler.java
package io.contract_testing.contractcase.configuration;

import java.util.Map;

/**
 * User-provided state setup and teardown functions.
 */
public class StateHandler {

  private final SetupFunction setupFn;
  private final TeardownFunction teardownFn;

  /**
   * Creates a state handler that has only a setup function which returns variables.
   *
   * @param setupFunction The setup function
   * @return A state handler for use in configuration
   */
  public static StateHandler setupFunction(SetupFunction setupFunction) {
    return new StateHandler(setupFunction, () -> {
    });
  }

  /**
   * Creates a state handler that has only a setup function and returns no variables.
   *
   * @param setupFunction The setup function with no variables
   * @return A new state handler
   */
  public static StateHandler setupFunction(SetupFunctionWithoutVariables setupFunction) {
    return new StateHandler(
        () -> {
          setupFunction.setup();
          // We use null to tell the core that there were no variables returned
          return null;
        },
        () -> {
        }
    );
  }

  /**
   * Creates a state handler that has both a setup function which returns variables, and a teardown
   * function.
   *
   * @param setupFunction    The setup function
   * @param teardownFunction the teardown function
   * @return The state handler
   */
  public static StateHandler setupAndTeardown(SetupFunction setupFunction,
      TeardownFunction teardownFunction) {
    return new StateHandler(setupFunction, teardownFunction);
  }

  /**
   * Creates a state handler that has both a setup function with no variables returned, and a
   * teardown function.
   *
   * @param setupFunction    The setup function
   * @param teardownFunction the teardown function
   * @return The state handler
   */
  public static StateHandler setupAndTeardown(SetupFunctionWithoutVariables setupFunction,
      TeardownFunction teardownFunction) {
    return new StateHandler(() -> {
      setupFunction.setup();
      // We use null to tell the core that there were no variables returned
      return null;
    }, teardownFunction);
  }

  private StateHandler(SetupFunction setupFn, TeardownFunction teardownFn) {
    this.setupFn = setupFn;
    this.teardownFn = teardownFn;
  }

  /**
   * Calls the setup function of this state handler. This is called by the framework, you don't need
   * to call it yourself.
   *
   * @return A map of state variables from the setup.
   */
  public Map<String, Object> setup() {
    return this.setupFn.setup();
  }

  /**
   * Calls the teardown function of this state handler. This is called by the framework, you don't
   * need to call it yourself.
   */
  public void teardown() {
    this.teardownFn.teardown();
  }
}
